sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -2092,11 +2098,247 @@ def set_application(self, channel, appl_code):
+        '''
+        self.byte_read_modify_write(consts.ADAPTIVE_INPUT_EQ_ENABLE_TX, host_lanes_mask, si_settings)
+
+    def stage_cdr_tx(self, host_lanes_mask, si_settings):
+        '''
+        This function applies TX CDR si settings
+        '''
+        self.byte_read_modify_write(consts.CDR_ENABLE_TX, host_lanes_mask, si_settings)
+
+    def stage_cdr_rx(self, host_lanes_mask, si_settings):
+        '''
+        This function applies RX CDR si settings
+        '''
+        self.byte_read_modify_write(consts.CDR_ENABLE_RX, host_lanes_mask, si_settings)
+
+    def stage_rx_si_settings(self, host_lanes_mask, si_settings):
+        for si_keys in si_settings:
+            if si_keys == consts.OUTPUT_EQ_PRE_CURSOR_TARGET_RX and self.get_rx_output_eq_pre_ctrl_supported():
+                self.stage_output_eq_pre_cursor_target_rx(host_lanes_mask, si_settings)
+            elif si_keys == consts.OUTPUT_EQ_POST_CURSOR_TARGET_RX and self.get_rx_output_eq_post_ctrl_supported():
+                self.stage_output_eq_post_cursor_target_rx(host_lanes_mask, si_settings)
+            elif si_keys == consts.OUTPUT_AMPLITUDE_TARGET_RX and self.get_rx_ouput_amp_ctrl_supported():
+                self.stage_output_amp_target_rx(host_lanes_mask, si_settings)
+            elif si_keys == consts.CDR_ENABLE_RX and self.get_rx_cdr_supported():
+                self.stage_cdr_tx(host_lanes_mask, si_settings)
+
+    def stage_tx_si_settings(self, host_lanes_mask, si_settings):
+        for si_keys in si_settings:
+            if si_keys == consts.FIXED_INPUT_EQ_TARGET_TX and self.get_tx_input_eq_fixed_supported():
+                self.stage_fixed_input_target_tx(host_lanes_mask, si_settings)
+            elif (si_keys == consts.ADAPTIVE_INPUT_EQ_RECALLED_TX and
+                 (self.get_tx_input_recall_buf1_supported() or self.get_tx_input_recall_buf2_supported())):
+                self.stage_adaptive_input_recall_tx(host_lanes_mask, si_settings)
+            elif si_keys == consts.ADAPTIVE_INPUT_EQ_ENABLE_TX and self.get_tx_input_adaptive_eq_supported():
+                self.stage_adaptive_input_eq_enable_tx(host_lanes_mask, si_settings)
+            elif si_keys == consts.CDR_ENABLE_TX and self.get_tx_cdr_supported():
+                self.stage_cdr_rx(host_lanes_mask, si_settings)
+
+    def stage_custom_si_settings(self, host_lanes_mask, optics_si_dict):
+        # Read and cache the existing SCS0 TX CTRL data
+        si_settings = self.xcvr_eeprom.read(consts.STAGED_CTRL0_TX_CTRL_FIELD)
+        if si_settings is None:
+            return None
+
+        # Replace the new values with cached SI values
+        for si_keys in optics_si_dict:
+            if si_keys in si_settings:
+                si_settings[si_keys] = optics_si_dict[si_keys]
+
+        # stage RX si settings
+        self.stage_rx_si_settings(host_lanes_mask, si_settings)
+
+        #stage TX si settings
+        self.stage_tx_si_settings(host_lanes_mask, si_settings)
```

Comment:
return status to check failure and return failure to caller?

Reply:
taken care